extensions/PhabBugz/lib/Feed.pm

```diff
@@ -0,0 +1,306 @@
+        return;
+    }
+
+    $self->logger->info("FEED: Fetching new transactions");
+
+    my $last_ts = $dbh->selectrow_array("
+        SELECT value FROM phabbugz WHERE name = 'feed_last_ts'");
+    $last_ts ||= 0;
+    $self->logger->debug("QUERY LAST_TS: $last_ts");
+
+    # Check for new transctions (stories)
+    my $transactions = $self->feed_transactions($last_ts);
+    if (!%$transactions) {
+        $self->logger->info("FEED: No new transactions");
+        return;
+    }
+
+    # Process each story
+    foreach my $story (keys %$transactions) {
+        my $skip = 0;
+        my $story_data  = $transactions->{$story};
+        my $author_phid = $story_data->{authorPHID};
+        my $object_phid = $story_data->{objectPHID};
+        my $story_text  = $story_data->{text};
+        my $story_epoch = $story_data->{epoch};
+
+        $self->logger->debug("STORY PHID: $story");
+        $self->logger->debug("STORY_EPOCH: $story_epoch");
+        $self->logger->debug("AUTHOR PHID: $author_phid");
+        $self->logger->debug("OBJECT PHID: $object_phid");
+        $self->logger->debug("STORY TEXT: $story_text");
+
+        # Only interested in changes to revisions for now.
+        if ($object_phid !~ /^PHID-DREV/) {
+            $self->logger->debug("SKIP: Not a revision change");
+            $skip = 1;
+        }
+
+        # Skip changes done by phab-bot user
+        my $phab_users = get_phab_bmo_ids({ phids => [$author_phid] });
+        if (!$skip && @$phab_users) {
+            my $user = Bugzilla::User->new({ id => $phab_users->[0]->{id}, cache => 1 });
+            $skip = 1 if $user->login eq PHAB_AUTOMATION_USER;
+        }
+
+        if (!$skip) {
+            my $revision = Bugzilla::Extension::PhabBugz::Revision->new({ phids => [$object_phid] });
+            $self->process_revision_change($revision, $story_text);
+        }
+        else {
+            $self->logger->info('SKIPPING');
+        }
+
+        # Store the largest last epoch + 1 so we can start from there in the next session
+        $story_epoch++;
+        $self->logger->debug("UPDATING LAST_TS: $story_epoch");
+        $dbh->do("REPLACE INTO phabbugz (name, value) VALUES ('feed_last_ts', ?)",
+                 undef, $story_epoch);
+    }
+}
+
+sub process_revision_change {
+    my ($self, $revision, $story_text) = @_;
+
+    # Pre setup before making changes
+    my $old_user = set_phab_user();
+
+    my $is_shadow_db = Bugzilla->is_shadow_db;
+    Bugzilla->switch_to_main_db if $is_shadow_db;
+
+    my $dbh = Bugzilla->dbh;
+    $dbh->bz_start_transaction;
+
+    my ($timestamp) = Bugzilla->dbh->selectrow_array("SELECT NOW()");
+
+    my $log_message = sprintf(
+        "REVISION CHANGE FOUND: D%d: %s | bug: %d | %s",
+        $revision->id,
+        $revision->title,
+        $revision->bug_id,
+        $story_text);
+    $self->logger->info($log_message);
+
+    my $bug = Bugzilla::Bug->new($revision->bug_id);
```

Comment:
It is possible for phabricator to initiate a writable SQL injection attack against BMO.
Two things -- the {fields} returned from phabricator should be validated, and the long form ->new({id => $revison->bug_id, cache => 1 }) should be used here